lib/agent.js

```diff
@@ -24,53 +25,99 @@ class Agent extends EventEmitter {
     }
 
     this[kFactory] = (origin) => factory(origin, opts)
-    this[kCache] = new Map()
+    this[kClients] = new Map()
 
     const agent = this
 
-    this[kOnConnect] = function onConnect (client) {
-      agent.emit('connect', client)
+    this[kOnConnect] = function onConnect (origin, clients) {
+      agent.emit('connect', origin, [...clients, agent])
     }
 
-    this[kOnDisconnect] = function onDestroy (client, err) {
+    this[kOnDrain] = function onDrain (origin) {
+      agent.emit('drain', origin)
+    }
+
+    this[kOnDisconnect] = function onDestroy (origin, clients, err) {
       if (this.connected === 0 && this.size === 0) {
         this.off('disconnect', agent[kOnDisconnect])
-        agent[kCache].delete(this.origin)
+        agent[kClients].delete(this.origin)
       }
 
-      agent.emit('disconnect', client, err)
+      agent.emit('disconnect', origin, [...clients, agent], err)
+    }
+  }
+
+  get connected () {
+    let ret = 0
+    for (const { connected } of this[kClients].values()) {
+      ret += connected
     }
+    return ret
   }
 
-  get (origin) {
-    if (typeof origin !== 'string' || origin === '') {
+  get size () {
+    let ret = 0
+    for (const { size } of this[kClients].values()) {
+      ret += size
+    }
+    return ret
+  }
+
+  get pending () {
+    let ret = 0
+    for (const { pending } of this[kClients].values()) {
+      ret += pending
+    }
+    return ret
+  }
+
+  get running () {
+    let ret = 0
+    for (const { running } of this[kClients].values()) {
+      ret += running
+    }
+    return ret
+  }
+
+  // TODO: get closed ()
+  // TODO: get destroyed ()
+
+  dispatch (opts, handler) {
+    if (!opts || typeof opts !== 'object') {
+      throw new InvalidArgumentError('opts must be an object.')
+    }
+
+    if (typeof opts.origin !== 'string' || opts.origin === '') {
       throw new InvalidArgumentError('Origin must be a non-empty string.')
     }
 
-    let pool = this[kCache].get(origin)
+    let pool = this[kClients].get(opts.origin)
 
     if (!pool) {
-      pool = this[kFactory](origin)
+      pool = this[kFactory](opts.origin)
```

Comment:
Not sure if this comment it's too premature, If we were building another Agent extending from this one, we would have to always add this manual bootstrap logic, which I think would be always the same for every custom Agent. That is what `get` provided, can we centralize it in some other `Agent.prototype` method?

Reply:
Reimplementing what agent does is rather trivial though?

Reply:
I would be happy to add some form of "protected" get later if the need ever arises. I'm a little skeptical whether it's actually necessary in practice.